hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -131,17 +182,25 @@ public List<R> consume(
 
 			final ResultHandler<R> resultHandlerToUse;
 
-			if ( uniqueSemantic == UniqueSemantic.ALLOW
-					&& domainResultJavaType instanceof EntityJavaType ) {
+			boolean isEnityResultType = domainResultJavaType instanceof EntityJavaType;
+			if ( uniqueSemantic == UniqueSemantic.ALLOW && isEnityResultType ) {
 				resultHandlerToUse = ListResultsConsumer::deDuplicationHandling;
 			}
 			else {
 				resultHandlerToUse = this.resultHandler;
 			}
 
+			final Results<R> results;
+			if ( isEnityResultType ) {
+				results = new EntityResult<>( domainResultJavaType );
+			}
+			else {
+				results = new Results<>( domainResultJavaType );
+			}
+
 			while ( rowProcessingState.next() ) {
 				final R row = rowReader.readRow( rowProcessingState, processingOptions );
-				resultHandlerToUse.handle( row, domainResultJavaType, results, rowProcessingState );
+				resultHandlerToUse.handle( row, results, rowProcessingState );
 				rowProcessingState.finishRowProcessing();
 			}
```

Comment:
Overall, I think I would prefer getting rid of `resultHandler` as I believe the megamorphic call (3 implementors) to `ResultHandler#handle` may be another bottleneck when processing big result lists, but that's just a thought. I understand that the following introduces a bit code duplication, but maybe you can try that and see if it improves the situation with the 1M rows even further.

```
		final UniqueSemantic uniqueSemantic = this.uniqueSemantic == UniqueSemantic.ALLOW && isEnityResultType ? UniqueSemantic.FILTER : this.uniqueSemantic;
		switch ( uniqueSemantic ) {
			case UniqueSemantic.FILTER: {
				while ( rowProcessingState.next() ) {
					final R row = rowReader.readRow( rowProcessingState, processingOptions );
					results.addUnique( result );
					rowProcessingState.finishRowProcessing();
				}
				break;
			}
			case UniqueSemantic.ASSERT: {
				while ( rowProcessingState.next() ) {
					final R row = rowReader.readRow( rowProcessingState, processingOptions );
					if ( !results.addUnique( result ) && !rowProcessingState.hasCollectionInitializers ) {
						throw new HibernateException(
							String.format(
									Locale.ROOT,
									"Duplicate row was found and `%s` was specified",
									UniqueSemantic.ASSERT
							)
						);
					}
					rowProcessingState.finishRowProcessing();
				}
				break;
			default: {
				while ( rowProcessingState.next() ) {
					final R row = rowReader.readRow( rowProcessingState, processingOptions );
					results.add( result );
					rowProcessingState.finishRowProcessing();
				}
				break;
			}
		}
```

Reply:
yes, if we could avoid the megamorphic calls here that would be great.

Reply:
Thanks! done :)